pandas/core/indexes/interval.py

```diff
@@ -766,8 +766,13 @@ def get_loc(self, key, method=None):
                 key = Interval(left, right, key.closed)
             else:
                 key = self._maybe_cast_slice_bound(key, 'left', None)
-
-            start, stop = self._find_non_overlapping_monotonic_bounds(key)
+            try:
```

Comment:
You'll also need to do something similar in the `else` branch to cover the overlapping/non-monotonic case, e.g. I think something like `pd.IntervalIndex.from_tuples([(1, 3), (2, 4), (0, 2)]).get_loc('foo')` will still fail.

Reply:
Maybe it should be the `engine` that should properly raise a KeyError? (eg the int64 engine does that)

Reply:
I still need to look into @jorisvandenbossche's comment on raising the error in the engine itself (especially if that's the behaviour for int64), but I think I've addressed everything else.

Reply:
@jorisvandenbossche : There is code in place within the `engine` that raises a `KeyError`, but strings queries fail before it gets there since the engine is expecting a `scalar_t` type (fused type consisting of numeric types) for `key`:
https://github.com/pandas-dev/pandas/blob/2e38d5552a5c7b2c0091cecddd483f4f08ad1d2c/pandas/_libs/intervaltree.pxi.in#L104

I'm not super well versed in Cython.  Is there a graceful way to force this to raise a `KeyError` within the Cython code?  Removing the `scalar_t` type gets a step further but still raises a `TypeError` as the code expects things to be comparable (probably some perf implications to removing it too).

Reply:
Yeah, the other engines have the key as `object` typed, and then afterwards do a check of that. 
But for me fine as well to leave that for now, and do the check here in the level above that. But on the long term would still be good to make the behaviour consistent throughout the different engines.